fedimint-core/src/api.rs

```diff
@@ -440,6 +474,78 @@ where
         .try_into_inner(decoders)
         .map_err(|e| anyhow!(e.to_string()))
     }
+}
+
+#[apply(async_trait_maybe_send!)]
+impl<T> IRawFederationApi for GlobalFederationApiWithCache<T>
+where
+    T: IRawFederationApi + MaybeSend + MaybeSync + 'static,
+{
+    fn all_peers(&self) -> &BTreeSet<PeerId> {
+        self.inner.all_peers()
+    }
+
+    fn with_module(&self, id: ModuleInstanceId) -> DynModuleApi {
+        self.inner.with_module(id)
+    }
+
+    /// Make request to a specific federation peer by `peer_id`
+    async fn request_raw(
+        &self,
+        peer_id: PeerId,
+        method: &str,
+        params: &[Value],
+    ) -> result::Result<Value, JsonRpcClientError> {
+        self.inner.request_raw(peer_id, method, params).await
+    }
+}
+
+#[apply(async_trait_maybe_send!)]
+impl<T> IGlobalFederationApi for GlobalFederationApiWithCache<T>
+where
+    T: IRawFederationApi + MaybeSend + MaybeSync + 'static,
+{
+    async fn await_block(
+        &self,
+        block_index: u64,
+        decoders: &ModuleDecoderRegistry,
+    ) -> anyhow::Result<SessionOutcome> {
+        let mut lru_lock = self.await_block_lru.lock().await;
+
+        let block_entry_arc = lru_lock
+            .get_or_insert(block_index, || Arc::new(tokio::sync::Mutex::new(None)))
+            .clone();
+
+        // we drop the lru lock so requests for other `block_index` can work in parallel
+        drop(lru_lock);
+
+        // but multiple request for the same `block_index` will serialize on the
+        // per-index lock, avoiding doing the same work
+        let mut block_lock = block_entry_arc.lock().await;
+
+        if let Some(cached_res) = block_lock.clone() {
+            debug!(block_index, "Using a cached await_block response");
+            return Ok(cached_res);
+        }
+
+        let block = self.await_block_raw(block_index, decoders).await?;
+
+        *block_lock = Some(block.clone());
```

Comment:
Should we also debug log wen we are fetching the block or is that already done by `await_block_raw`?

Reply:
Will check when I rebase, it should also be easier to review in full.

Reply:
Unclear if it's really needed.

Reply:
I mean, you do have the log when the block/session was cached. But fine by me :man_shrugging: more logging more good though

Reply:
there is already log in await_block_raw